### PR TITLE
KAFKA-8560. The Kafka protocol generator should support common structures

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Comment by cmccabe** [comment]
retest this please

**Comment by cmccabe** [comment]
retest this please

File: clients/src/main/resources/common/message/LeaderAndIsrRequest.json [hunk 1]
```diff
@@ -28,59 +28,48 @@
...
-        { "name": "IsrReplicas", "type": "[]int32", "versions": "0+",
-          "about": "The in-sync replica IDs." },
-        { "name": "ZkVersion", "type": "int32", "versions": "0+",
-          "about": "The ZooKeeper version." },
-        { "name": "Replicas", "type": "[]int32", "versions": "0+",
-          "about": "The replica IDs." },
-        { "name": "IsNew", "type": "bool", "versions": "1+", "default": "false", "ignorable": true, 
-          "about": "Whether the replica should have existed on the broker or not." }
-      ]}
+      { "name": "PartitionStatesV0", "type": "[]LeaderAndIsrRequestPartition", "versions": "2+",
```
Comment (jsancio) [inline clients/src/main/resources/common/message/LeaderAndIsrRequest.json:40]: Looks like we are changing the `name` and the `versions`. Is this a bug in the old schema?

Comment (gwenshap) [reply clients/src/main/resources/common/message/LeaderAndIsrRequest.json:40]: Not sure if exactly a bug, but the old schema was definitely more confusing.  `PartitionState` itself is a field in `TopicStates` and since `TopicState` is for versions 2+, this field wouldn't show up in older versions no matter what is the specific version of the field.

Colin unified this field into the new common structure `LeaderAndIsrRequestPartition` which has `TopicName` that only shows up in versions 0-1. So if you have the 2+ `TopicStates` structure, this field won't show up. But when used in  the older (0-1) struct `PartitionStatesV0`, it will be included. This is basically what the new comment above is trying to explain.

File: generator/src/main/java/org/apache/kafka/message/StructRegistry.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * Contains structure data for Kafka MessageData classes.
+ */
+final class StructRegistry {
+    private final Map<String, StructSpec> structSpecs;
+    private final Set<String> commonStructNames;
```
Comment (gwenshap) [inline generator/src/main/java/org/apache/kafka/message/StructRegistry.java]: Why do you need both a map and a set? TreeMap gives you the key set and key iterator for "free"?

**Comment by gwenshap** [comment]
Looks good. Reviewing would have been easier if the PR included few sentences on the struct registry and why common structs and normal structs are handled separately in generation. Took me some back-and-forth to figure it out. 

Are there future plan / need to share common structs between message classes? reuse seems a bit limited right now.

Can you add a test-case to cover the expected errors? (common struct with fields, duplicate common structs, etc)

File: clients/src/main/resources/common/message/LeaderAndIsrRequest.json (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/resources/common/message/LeaderAndIsrRequest.json:40]: Thanks for the explanation. That's right.

**Comment by cmccabe** [comment]
> Looks good. Reviewing would have been easier if the PR included few sentences on the struct registry and why common structs and normal structs are handled separately in generation. Took me some back-and-forth to figure it out.

Thanks for the review.  Just to give a little more background on this, the main rationale behind the `StructRegistry` is that we want to make a first pass over the data so that we can know ahead of time which structs exist.  That way, if we hit a typo, we can give an error on the line when it happens.  Another reason is because sometimes we have to know something about the struct, like if it should be serialized as an ImplicitLinkedHashCollection or not.  We can't really know that in a one-pass system where the struct hasn't been processed yet.

> Are there future plan / need to share common structs between message classes? reuse seems a bit limited right now.

The big problem is that the versions are unlikely to line up between different message types.  So the obvious way of doing it wouldn't really work that well.  In general, we'd like to be able to add something to certain version of message type X without also adding it to Y, and cross-API structures make that hard (or impossible?)

In practice, if we want to share stuff between messages, we can do it by creating a new field type.  For example, we might have a UUID field type if we ever end up using UUIDs.  We could have some kind of Error type which mapped out to `org.apache.kafka.common.protocol.Errors` or to `ApiError` without the need for manual fiddling.

We could also may be do something with interfaces, where if you had two very similar structures in two RPCs, they could both implement the same Java interface, so you could have common code to handle them in many cases (if that made sense...)

As it is, this change fixes a current problem, which is that sometimes we reorganize things within a single message, and we want to avoid a lot of duplication.  I expect we'll see more of these cases as time goes on, as well.

> Can you add a test-case to cover the expected errors? (common struct with fields, duplicate common structs, etc)

Will add

**Comment by gwenshap** [comment]
LGTM. Merging.

**Comment by ijuma** [comment]
Thanks for the PR, this is useful. There is a very useful discussion in the PR, but the commit message and PR description are blank. It would be helpful if the PR description and commit message included all the information one would need to understand the change (in the future).
